src/queue.rs

```diff
@@ -24,69 +22,53 @@ use std::sync::mpsc::{channel, Receiver, Sender};
 ///   a new sound.
 /// - If you pass `false`, then the queue will report that it has finished playing.
 ///
-pub fn queue<S>(keep_alive_if_empty: bool) -> (Arc<SourcesQueueInput<S>>, SourcesQueueOutput<S>)
+pub fn queue<S>(keep_alive_if_empty: bool) -> (Arc<QueueControls<S>>, QueueSource<S>)
 where
     S: Sample + Send + 'static,
 {
-    let input = Arc::new(SourcesQueueInput {
-        next_sounds: Mutex::new(Vec::new()),
+    let input = Arc::new(QueueControls {
+        next_sounds: Mutex::new(VecDeque::new()),
         keep_alive_if_empty: AtomicBool::new(keep_alive_if_empty),
     });
 
-    let output = SourcesQueueOutput {
+    let output = QueueSource {
         current: Box::new(Empty::<S>::new()) as Box<_>,
-        signal_after_end: None,
         input: input.clone(),
+        filling_initial_silence: Cell::new(false),
+        next_sample: None,
     };
 
     (input, output)
 }
 
-// TODO: consider reimplementing this with `from_factory`
-
 type Sound<S> = Box<dyn Source<Item = S> + Send>;
-type SignalDone = Option<Sender<()>>;
-
 /// The input of the queue.
-pub struct SourcesQueueInput<S> {
-    next_sounds: Mutex<Vec<(Sound<S>, SignalDone)>>,
+pub struct QueueControls<S> {
```

Comment:
Maybe just "Queue"? This would be more in line with other control/source struct pairs.

Reply:
done (in #706)